src/Control/Immortal.hs

```diff
@@ -124,24 +114,24 @@ threadId (Thread pid _ _) = pid
 --
 -- This is nothing more than a simple wrapper around 'try'.
 onFinish
-  :: MonadBaseControl IO m
+  :: MonadUnliftIO m
   => (Either SomeException () -> m ())
   -> m () -> m ()
-onFinish cb a = try a >>= cb
+onFinish cb a = withRunInIO $ \run -> try (run a) >>= run . cb
 
 -- | Like 'onFinish', but the callback does not run when the thread is
 -- mortalized (i.e. when the exit is expected).
 --
 -- The 'Thread' argument is used to find out the mortality of the thread.
 onUnexpectedFinish
-  :: MonadBaseControl IO m
+  :: MonadUnliftIO m
   => Thread
   -> (Either SomeException () -> m ())
   -> m ()
   -> m ()
-onUnexpectedFinish (Thread _ stopRef _) cb a = do
-  r <- try a
-  expected <- liftBase $ readIORef stopRef
+onUnexpectedFinish (Thread _ stopRef _) cb a = withRunInIO $ \run -> do
+  r <- try $ run a
```

Comment:
Wouldn't it be cleaner to depend on unlift and use the unlifted functions?

I genuinely don't know; how much here could be simplified with unlift?

Reply:
Depending on `unlift` would provide a (rougly) equivalent `try`, `mask`, etc. but -

1. It was easy enough to write with `unliftio-core`
2. `unliftio` would bring in a few extra dependencies like `directory`, `filepath`, `process`, `unix`, so I decided to try to get away without it
3. The semantics would technically change a bit as the unlifted functions ignore asynchronous exceptions by default. The backgrounded thread could then be brought down by a `killThread` without being `mortalized` first, for example, so the logic would have to be tidied up a bit.